solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformer.java

```diff
@@ -123,6 +124,16 @@ public void transform(SolrDocument rootDoc, int rootDocId) {
 
         // Do we need to do anything with this doc (either ancestor or matched the child query)
         if (isAncestor || childDocSet == null || childDocSet.exists(docId)) {
+
+          if(limit != -1) {
+            if(!isAncestor) {
+              if(matches == limit) {
+                continue;
+              }
+              ++matches;
```

Comment:
I think matches should be incremented if it's in childDocSet (includes childDocSet being null).  Wether it's an ancestor or not doesn't matter I think.  You could pull out a new variable isInChildDocSet.  Or I suppose simply consider all a match; what I see what you just did as I write this; that's fine too.